### PR TITLE
Generalize QSU dsl

### LINKED ISSUE #3077: Specification for ApplyProvenance
Write the initial specification for `ApplyProvenance`.

### PR DESCRIPTION
**Pull request by alissapajer**
Adds first test for `ApplyProvenance`. 

Uses `EnvT` for pattern functor for `QSUGraph`.

Closes #3077

### CODE REVIEW

File: connector/src/test/scala/quasar/qscript/qsu/ApplyProvenanceSpec.scala [hunk 1]
```diff
@@ -16,17 +16,86 @@
...
+	val actual: PlannerError \/ AuthenticatedQSU[Fix] = app[F](inputGraph)
+
+	// TODO use Show
+        actual.bimap[MatchResult[S], MatchResult[S]](
+        { err =>
+          failure(s"provenance application produced unexpected planner error: ${err}", s)
+        },
+        { case auth @ AuthenticatedQSU(resultGraph, resultDims) =>
+          result(
+            (resultGraph == inputGraph) && (resultDims == expectedDims), // TODO better equals
```
Comment (alissapajer) [inline connector/src/test/scala/quasar/qscript/qsu/ApplyProvenanceSpec.scala]: Going to be using `EqMap` here instead.

**Comment by djspiewak** [comment]
@alissapajer I'm a little confused I think…  Why is it that `Dsl` has to operate in a functor now?  I probably missed it in the diff somewhere.

File: connector/src/test/scala/quasar/qscript/qsu/ApplyProvenanceSpec.scala [hunk 2]
```diff
@@ -16,17 +16,86 @@
...
 
   "provenance application" should {
-    "produce empty provenance for constants" in prop { data: Data =>
-      ok
+
+    "produce provenance for map" in {
+
+      val fm: FreeMap = func.Add(HoleF, IntLit(17))
+
+      val tree: Cofree[QSU, String] =
```
Comment (alissapajer) [inline connector/src/test/scala/quasar/qscript/qsu/ApplyProvenanceSpec.scala]: @djspiewak the goal is to use `AnnotatedDsl` to create a `Cofree[QSU, String]` for testing.

Comment (djspiewak) [reply connector/src/test/scala/quasar/qscript/qsu/ApplyProvenanceSpec.scala]: Gotcha

File: connector/src/main/scala/quasar/qscript/qsu/QScriptUniform.scala [hunk 3]
```diff
@@ -355,6 +368,11 @@ object QScriptUniform {
         case Subset(f, op, c) => (f, op, c)
       } { case (f, op, c) => Subset(f, op, c) }
 
+    def subsetEnvT[A]: Prism[(Symbol, QScriptUniform[A]), (Symbol, (A, SelectionOp, A))] =
```
Comment (wemrysi) [inline connector/src/main/scala/quasar/qscript/qsu/QScriptUniform.scala]: Is this needed?

File: connector/src/test/scala/quasar/qscript/qsu/ApplyProvenanceSpec.scala [hunk 4]
```diff
@@ -16,17 +16,86 @@
...
+        qsu.map("name0",
```
Comment (wemrysi) [inline connector/src/test/scala/quasar/qscript/qsu/ApplyProvenanceSpec.scala]: Nice, that works out rather cleanly!

File: connector/src/main/scala/quasar/qscript/qsu/QScriptUniform.scala (hunk 3 above)
Comment (alissapajer) [reply connector/src/main/scala/quasar/qscript/qsu/QScriptUniform.scala]: Oh, this is left over from when I was implementing it. I'll delete.

**Comment by alissapajer** [comment]
@wemrysi this should be ready to go!

**Pull request merged**